File: NeetCode/Linked_List/Copy_Linked_List_with_Random_Pointer/copy__linked__list_with__random__pointer.py
from typing import Optional, List
from common.listnode import ListNode, build_list, print_list


class Node:
    def __init__(self, x: int, next: "Node" = None, random: "Node" = None):
        self.val = int(x)
        self.next = next
        self.random = random


class Solution:
    def copyRandomList(self, head: "Optional[Node]") -> "Optional[Node]":
        if head == None:
            return None

        nodeDict = {}

        ptr = head
        last = None
        while ptr:
            tmp = Node(ptr.val)
            nodeDict[ptr] = tmp
            ptr = ptr.next

        ptr = head
        while ptr:
            nodeDict[ptr].next = nodeDict.get(ptr.next)
            nodeDict[ptr].random = nodeDict.get(ptr.random)
            ptr = ptr.next

        return nodeDict[head]


if __name__ == "__main__":
    sol = Solution()
    print("Running Solution...")

File: NeetCode/Linked_List/Copy_Linked_List_with_Random_Pointer/Copy_Linked_List_with_Random_Pointer.cpp
// Copy_Linked_List_with_Random_Pointer.cpp
#include <algorithm>
#include <array>
#include <cassert>
#include <cctype>
#include <climits>
#include <cmath>
#include <deque>
#include <functional>
#include <iostream>
#include <map>
#include <numeric>
#include <queue>
#include <set>
#include <stack>
#include <string>
#include <unordered_map>
#include <unordered_set>
#include <utility>
#include <vector>

using namespace std;
class Node
{
public:
    int val;
    Node *next;
    Node *random;

    Node(int _val)
    {
        val = _val;
        next = NULL;
        random = NULL;
    }
};
class Solution
{
public:
    Node *copyRandomList(Node *head)
    {
        unordered_map<Node *, Node *> oldToNew;

        Node *ptr = head, *tmp;
        while (ptr)
        {
            tmp = new Node(ptr->val);
            oldToNew[ptr] = tmp;
            ptr = ptr->next;
        }

        ptr = head;

        while (ptr)
        {
            oldToNew[ptr]->next = oldToNew[ptr->next];
            oldToNew[ptr]->random = oldToNew[ptr->random];
            ptr = ptr->next;
        }
        return oldToNew[head];
    }
};
